aten/src/ATen/native/TensorIndexing.cpp

```diff
@@ -154,5 +152,189 @@ std::ostream& operator<<(std::ostream& stream, const std::vector<TensorIndex>& t
   return stream;
 }
 
+// This mirrors `count_specified_dimensions` in torch/csrc/autograd/python_variable_indexing.cpp
+int64_t count_specified_dimensions(ArrayRef<TensorIndex> indices) {
+  // Count the number of indexed dimensions (everything but ellipsis and None)
+  int64_t count = 0;
+  size_t size = indices.size();
+  for (size_t i = 0; i < size; i++) {
+    auto& obj = indices[i];
+    if (obj.is_tensor()) {
+      auto& tensor = obj.tensor();
+      if (tensor.scalar_type() == kByte || tensor.scalar_type() == kBool) {
+        count += tensor.dim();
+      } else {
+        count++;
+      }
+    } else if (!obj.is_none() && !obj.is_ellipsis() && !obj.is_boolean()) {
+      count++;
+    }
+  }
+  return count;
+}
+
+// This mirrors `applySlicing` in torch/csrc/autograd/python_variable_indexing.cpp
+Tensor applySlicing(const Tensor& self, ArrayRef<TensorIndex> indices, std::vector<Tensor>& outIndices) {
+  int64_t size = indices.size();
+  int64_t dim = 0;
+  int64_t specified_dims = count_specified_dimensions(indices);
+
+  TORCH_CHECK_INDEX(specified_dims <= self.dim(), "too many indices for tensor of dimension ", (int)self.dim());
+
+  Tensor result = self;
+  for (int64_t i = 0; i < size; i++) {
+    auto& obj = indices[i];
+    if (obj.is_integer()) {
+      result = handleInteger(result, dim, obj.integer(), i);
```

Comment:
I am comparing the code here to the old code, and I am thinking to myself, "The old code was way clearer; these 'helper' functions are really small and used to be inlined, and having them be inlined was a lot clearer." So then I think, "Why are these helper functions out-of-line, as opposed to inline?" And the answer is because they're called twice: once from this function, and once from the Python loop. But the claim I was making in https://godbolt.org/z/8ruXwy is that you can put the *entire* if-statement in a single helper function (rather than breaking it into all of these pieces), write the conversion from PyObject to TensorIndex as one function, and then just call that helper function. To me, that seems like it would be clearer.

Reply:
I updated the code to use the "put the entire if-statement in a single helper function, write the conversion from PyObject to TensorIndex as one function, and then call that helper function" design, which makes the code a lot cleaner.

I also compared two approaches with regard to optimizing away the TensorIndex object creation in the Python path: 1. Pass the TensorIndex object to a handler function immediately after object creation, and let the compiler do the optimization; 2. Have a global temporary TensorIndex object, and always reuse that object (with its value updated accordingly) when we need to pass Python index info into C++. The current form of the PR takes the second approach, because I think the success of the first approach is dependent on how the compiler does the optimization, and it's not a guaranteed thing, while in the second approach it's more certain that no new TensorIndex object would be created. We also know that there won't be concurrent access to the global temporary TensorIndex object since Python doesn't have multithreading, so there won't be data access race condition.